app/src/main/java/tech/relaycorp/letro/awala/AwalaManager.kt

```diff
@@ -117,6 +98,33 @@ class AwalaManagerImpl @Inject constructor(
         }
     }
 
+    private suspend fun loadFirstPartyEndpoint(): FirstPartyEndpoint {
+        val firstPartyEndpointNodeId = awalaRepository.getServerFirstPartyEndpointNodeId()
+            ?: registerFirstPartyEndpointIfNeeded()?.nodeId
+            ?: throw IllegalStateException("You should register first party endpoint first!")
+        return firstPartyEndpoint ?: loadNonNullFirstPartyEndpoint(firstPartyEndpointNodeId)
+    }
+
+    private suspend fun loadThirdPartyEndpoint(recipient: MessageRecipient): ThirdPartyEndpoint {
+        if (recipient is MessageRecipient.Server) {
+            thirdPartyServerEndpoint?.let {
+                return it
+            }
+        }
+        val thirdPartyEndpointNodeId = when (recipient) {
+            is MessageRecipient.Server -> {
+                recipient.nodeId
+                    ?: awalaRepository.getServerThirdPartyEndpointNodeId()
+                    ?: importServerThirdPartyEndpointIfNeeded()?.nodeId
+                    ?: throw IllegalStateException("You should register third party endpoint first!")
+            }
+            else -> {
+                throw IllegalStateException("User messages are not supported yet!")
```

Comment:
Even once we implement messaging, this logic would still be useful: We should still get a request to load a non-existing third-party endpoint.

However, I think this logic should be more generic. For example:

```suggestion
                // Log the missing endpoint id, but don't include it in the exception to avoid leaking it
                // to the Google Play Store Developer Console (which would be a privacy violation).
                Log.e("Cannot find third-party endpoint ${recipient.nodeId}")
                throw IllegalStateException("Cannot find third-party endpoint")
```

Reply:
That's done